cmds/webboot/webboot.go

```diff
@@ -66,6 +59,52 @@ func (d *DownloadByLink) Exec() error {
 	return nil
 }
 
+// Exec displays subdirectory or cached isos under the path directory
+func (g *DirGroup) Exec() error {
+	entries := []menu.Entry{}
+	readerInfos, err := ioutil.ReadDir(g.path)
+	if err != nil {
+		return err
+	}
+
+	// check the directory, if there is a subdirectory, add another DirGroup option
+	// if there is iso file, add a cachedISO option
+	for _, info := range readerInfos {
+		if info.IsDir() {
+			entries = append(entries, &DirGroup{
+				label:    info.Name(),
+				path:     filepath.Join(g.path, info.Name()),
+				uiEvents: g.uiEvents,
+			})
+		} else if filepath.Ext(info.Name()) == ".iso" {
+			iso := &CachedISO{
+				path:  filepath.Join(g.path, info.Name()),
+				label: info.Name(),
+			}
+			entries = append(entries, iso)
+		}
+	}
+	_, err = menu.DisplayMenu("Distros", "Choose an option", entries, g.uiEvents)
+	return err
+}
+
+// Exec displays the next level of bookmarks
+func (g *BookmarkGroup) Exec() error {
+	nextLevel := []menu.Entry{}
+	for _, e := range g.entries {
+		if b, ok := e.(*BookmarkGroup); ok {
+			b.uiEvents = g.uiEvents
+			nextLevel = append(nextLevel, b)
+		} else {
+			nextLevel = append(nextLevel, e)
+		}
+
+	}
+	_, err := menu.DisplayMenu("Bookmarks", "Choose an option", nextLevel, g.uiEvents)
```

Comment:
I'm thinking we might be able to avoid having to add all these ui.Events in the structs. The DisplayMenu function returns the entry that was chosen, and if you break up the Exec function to call a helper function that takes in a ui.Events and returns the result of the next menu, would that helper function then give you everything you need for a test?